DigitalLearningSolutions.Web/Controllers/LoginController.cs

```diff
@@ -57,21 +57,30 @@ public IActionResult Index(LoginViewModel model)
                 return View("Index", model);
             }
 
+            if (verifiedAdminUser != null && verifiedDelegateUsers.Count == 0)
+            {
+                verifiedDelegateUsers =
+                    loginService.GetVerifiedDelegateUsersAssociatedWithAdminUser(verifiedAdminUser, model.Password);

```

Comment:
Hmm.. I feel like we should get the associated delegates when we get the users in the first place, because:
* There is no reason to wait until we get the associated delegate accounts and if we get them early, the code above will deal with verifying the password. It's a bit different for finding an admin account associated with an unapproved delegate, because we need to handle that case differently from the UX perspective.
* It will keep the controller less crowded and make testing the code cleaner.

Also, we should only get the associated admin user if the delegate user is approved (moving that bit of logic up here means that if someone logs in with a delegate id that is for an unapproved account, we won't show them the "unapproved" page if they have an admin account associated with the email address of the delegate account).

Reply:
That all makes sense, but I am a bit confused about the final bit with approval. I appreciate that if a user logs in with a delegate ID for an unapproved account, they should be shown the "Account Not Approved" page. But how will this work when we get the Choose A Centre stuff working? On there, I imagine if they log in with an unapproved delegate Id, they still want to see all of their roles on all centres, creating a different set of behaviours if they have multiple centres or not (I don't know if we really need to be concerned with it on this bug fix or not, but seeing as I am currently working on the Choose a Centre ticket I might as well ask)

For clarity:
- Imagine a situation where a user has 3 accounts: An unapproved delegate at Centre 1, an Admin account also at Centre 1, and an approved delegate at Centre 2 that will not be logged in to.
- As they have multiple centres, when they log in they would see "Delegate" at Centre 2, and "Delegate and Admin" at Centre 1. If they choose to log in to Centre 1, what should happen? I think they should get logged in with both delegate and admin accounts, thus avoiding the Account Not Approved page, despite the fact that the delegate account is not approved. They get around this because they know they have an Admin account.
- If they did not have that other account at Centre 2, they would only be detected as at Centre 1. Therefore, if they log in with the delegate Id, the admin account will not be picked up before the approval check, and thus will be shown the Account Not Approved page.

This would be different behaviours based when logging in, based on a separate account to the one they are trying to access. What are our thoughts on this?

Reply:
The reason we show the unapproved page is to keep away users that get confused about the fact that they cannot log into their delegate account. Hence:
* If the user logs in with the unapproved delegate id (rather than their usual login that they're used to using with other accounts), we should always show unapproved page to communicate that the account is not approved. The main reason they'd be using this login is to access the new delegate account (this is different to the account they're used to using).
* If the user logs in with a different login (email/admin login/or approved delegate id), we should display the Choose a centre page with only approved accounts - so in your example it would be "Centre 1 - admin" and "Centre 2 - delegate".

Reply:
Right, I think I get it now: If a user with an unapproved delegate account tries to log in with that delegate id, they will always be shown the unapproved page, regardless as to the number of centres available to their email address (which is tied to their account, but not used to log in in this case). And that should be the only way to "log in" to an unapproved account, as using the email address would return other centres and not show the unapproved ones.

Reply:
Yes, great summary! 👍